src/dotty/tools/dotc/transform/Space.scala

```diff
@@ -0,0 +1,316 @@
+      else a
+    case (Typ(t1), Kon(t2, ss)) =>
+      if (isSubType(t1, t2)) minus(Kon(t2, getSignature(t2).map(Typ)), b)
+      else if (isSubType(t2, t1) && canDecompose(t1))
+        minus(Or(getPartitions(t1).map(Typ)), b)
+      else a
+    case (_, Or(ss)) =>
+      ss.foldLeft(a)(minus)
+    case (Or(ss), _) =>
+      Or(ss.map(minus(_, b)))
+    case (Kon(t1, ss), Typ(t2)) =>
+      if (isSubType(t1, t2)) Empty
+      else if (simplify(a) == Empty) Empty
+      else if (isSubType(t2, t1) && canDecompose(t1))
+        minus(Or(getPartitions(t1).map(Typ)), b)
+      else a
+    case (Kon(t1, ss1), Kon(t2, ss2)) =>
+      if (!isEqualType(t1, t2)) a
+      else if (ss1.zip(ss2).exists(p => simplify(intersect(p._1, p._2)) == Empty)) a
+      else if (ss1.zip(ss2).forall((subspace _).tupled)) Empty
+      else
+        Or(
+          ss1.zip(ss2).map((minus _).tupled).zip(0 to ss2.length - 1).map {
+            case (ri, i) => Kon(t1, ss1.updated(i, ri))
+          })
+  }
+}
+
+// scala implementation of space logic
+class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
+  def debug(s: String): Unit = {
+    if (ctx.debug) println(s)
+  }
+
+  // if roundUp, approximate extractors to its type
+  // else approximate extractors to Empty
+  def project(pat: tpd.Tree)(implicit ctx: Context, roundUp: Boolean = true): Space = pat match {
+    case Ident(_) => Typ(pat.tpe)
+    case Select(_, _) => Typ(pat.tpe)
+    case Alternative(trees) => Or(trees.map(project _))
+    case Bind(_, pat) => project(pat)
+    case UnApply(_, _, pats) =>
+      if (pat.tpe.classSymbol is Flags.CaseClass)
+        Kon(pat.tpe, pats.map(project _))
+      else if (roundUp) Typ(pat.tpe)
+      else Empty
+    case Typed(t@UnApply(_, _, _), _) => project(t)
+    case Typed(expr, _) => Typ(expr.tpe)
+    case _ =>
+      debug(s"========unkown tree: $pat========")
+      Empty
+  }
+
+  def strip(t: Type)(implicit ctx: Context): Type = t match {
+    case t: RefinedType => strip(t.underlying)
+    case _ => t
+  }
+
+  def isSubType(t1: Type, t2: Type): Boolean = {
+    val res = t1 <:< t2
+    debug(s"$t1 <: $t2 ? $res")
+    res
+  }
+
+  def isEqualType(t1: Type, t2: Type): Boolean = {
+    val res = t1 =:= t2
+    debug(s"$t1 == $t2 ? $res")
+    res
+  }
+
+  def getSignature(t: Type): List[Type] = {
+    val ktor = t.classSymbol.primaryConstructor.info
+
+    def typeParams(t: Type): List[Type] = t match {
+      case t: RefinedType => t.refinedInfo.asInstanceOf[TypeAlias].alias :: typeParams(t.parent)
+      case _ => Nil
+    }
+
+    val sign = (t match {
+      case t: RefinedType => ktor.appliedTo(typeParams(t))
+      case _ => ktor
+    }).firstParamTypes.map(_.stripTypeVar)
+
+    debug(s"====signature of $t: $sign")
+    sign
+  }
+
+  def getPartitions(t: Type): List[Type] = t match {
+    case OrType(t1, t2) => List(t1, t2)
+    case _ =>
+      val children = t.classSymbol.children.map { c =>
+        if (c is Flags.ModuleClass)
```

Comment:
I think it's nicer to import `Flags._` so that you can do `c.is(ModuleClass)` (also our current coding style discourage the use of infix notation, prefer `foo.is(bar)` to `foo is bar`)